### DISCUSSION TITLE
Writing to cache outside of updaters

### DISCUSSION THREAD

**Discussion by sarink** [urql-graphql/urql#2462] [Q&A] [answered]
I'm refactoring an app that uses apollo client to use urql. Because the cache can only be written to from an updater function, I'm not sure how to handle one of our use cases:

1. Client fetches items from our graphql api
1. Client launches third party iframe, iframe closes, in the close callback the client receives some updated item data to display in the UI optimistically
1. Behind the scenes, third party triggers a webhook in the api, which actually updates the items in the db
1. Now, the api sends a websocket message to the client to let it know the items are ready, and the client re-fetches them

I'm not sure how to implement step 2. The existing apollo client code is able to write the preview metadata to the cache directly in the react component that handles opening/closing the third party iframe, but urql explicitly disallows that

**Comment by sarink** [comment]
I've been reading a lot of other threads where this feature has been requested, and I understand urql's position: There should be no discrepancy in the data, the api should be the single and only source of truth for all data all the time.

In practice, however, when dealing with third party services and/or other non-graphql services... there _is_ sometimes a discrepancy in the data. In a pattern very familiar to urql, the UI can receive data optimistically from these other services _before_ the api has received it.

The alternatives seems to be:
1. Keeping this third party data in some other state management library, refactoring every component that uses it to both perform a query to get it from the api, _and_ check for it in our other state management library, then have some logic to decide which to prefer. This does not seem feasible, and it also breaks the "single source of truth" principle for our client (where the single source of truth is the graphcache)
2. Actually submitting this temporary placeholder data to the api in a mutation. This creates a race condition: if the official webhook came in first, the client's mutation would overwrite the "official" data with the "temporary" data. Perhaps this could be solved with timestamps and/or request ids? It is probably the most Correct solution, since it keeps the state totally in sync, but it's a lot of work and a brittle solution that could leave you in a _worse_ state if something went wrong, all that effort and risk for what benefit?
3. An api "echo mutation" resolver that returns what the client passed it. This requires a completely unnecessary round trip for the client, but it tricks urql into thinking the data actually came from the server (because it did, sort of), and is the simplest/most straightforward way to workaround urql's aggressive opinion on cache writes.

There are absolutely legitimate use cases for being able to access the cache directly. With no viable workarounds for a client to tell urql "plz pretend this data came from the server, it's legit, I promise", although I understand the design decision and opinion, I cannot agree with it. It looks like we'll either be resorting to ""solution"" 3, or abandoning this branch and going back to apollo :-/

Edit: Is there any way to implement a custom `fetchExchange` that would effectively implement number 3 above without _actually_ making a call to the server?

**Reply by kitten** [reply]
to be fair, what you're asking for here is a _temporary, optimistic_ change which is something that no GraphQL client supports externally as far as I know.
Hence, you'd be doing the exact same thing we want to avoid by not allowing this; having out-of-band updates that don't flow through GraphQL with a trigger that isn't in GraphQL, causing a cache update that is temporary with you having to roll it back 

Ultimately, this isn't 100% possible in the current description. Changes are either optimistic or they're not. We always have a concept of layers and rely on certain guarantees that GraphQL APIs implicitly make to keep the state true to the GraphQL API.
So, that's ultimately not making Graphcache the source of truth. The intention is to make your GraphQL API the source of truth.
In other words, this last part is a contract that when broken we don't cater for, because it's simply almost impossible to handle this complexity well manually.

However, I haven't yet heard of optimistic changes that are out-of-band and eventually are reflected by the API and I can actually see this become a viable alternative without allowing arbitrary updates, which inherently lead to footguns that are too easy to fire, so to speak.

I think though, I may need some more information on how this flow came about that you're describing, because there's some missing information on what triggers this iframe interaction, why it's necessary, and what ultimately leads to the update.
The reason I'm asking is because you will have race conditions here. It's unavoidable unless you ignore them. If data changes in the future and you're able to ask the GraphQL API for new data before these changes are applied you will assuredly see incorrect data. Answering what the intention here is is interesting to me because it may give us a path to provide safe out-of-band optimistic updates, which are inherently safer, because we can potentially block related queries from fetching while those are active 🤔

**Reply by sarink** [reply]
Hey @kitten , thanks for your fast response.

It's true that no client handles this case explicitly, but other clients (apollo) provide a `writeQuery` method that can be accessed from anywhere (like, an iframe's `onClose` callback), so out-of-band updates are not an issue in the first place.

Having used apollo for quite some time, I do agree with you that this design is a "footgun". However, I would also argue that both my case (which is a normal flow when integrating with oauth-like services), and [this one](https://github.com/FormidableLabs/urql-exchange-graphcache/issues/109) are good reasons for urql provide _some_ kind of escape hatch.

To be specific, we are using Plaid and their [link flow](https://plaid.com/docs/link/web/). When the plaid iframe closes, the [onSuccess](https://plaid.com/docs/link/web/#link-web-onsuccess-metadata) callback contains some of the updated `account` and `institution` metadata -- this is what we would like to display to the user. Plaid also triggers a series of webhooks on our api, which stores the updated data in our database, and when that is done processing the client will re-fetch this data from the api the normal way (overriding that initial "optimistic" data that the iframe callback gave us).

Whether removing, modifying, or linking a new plaid item, plaid gives us the necessary metadata to show the changes optimistically, and all of this optimistic data will be overridden by real server data at some point in the future.

Do these specifics help bring a little more clarity?

**Reply by sarink** [reply]
Just following up, we were able to do this by passing "extra" `variables` when performing the mutation, and utilizing that data in the `optimistic` and `updates` functions. https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#variables-for-optimistic-updates

**Comment by sarink** [comment, marked as answer]
Just following up, we were able to do this by passing "extra" `variables` when performing the mutation, and utilizing that data in the `optimistic` and `updates` functions. https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#variables-for-optimistic-updates
